File: source_code/java/Water.java
import java.util.List;
import java.util.ArrayList;

public class Water implements State {
  public int x, y;

  public Water(int x, int y) {
    this.x = x;
    this.y = y;
  }

  public String toString() {
    return "[" + x + "," + y + "]";
  }

  public Boolean isGoal() {
    if (y == 4)
      return true;
    return false;
  }

  public List<Successor> successors() {
    List<Successor> l = new ArrayList<Successor>();
    if (x > 0)
      l.add(new Successor(new Water(0, y), "Empty {3}", x));
    if (y > 0)
      l.add(new Successor(new Water(x, 0), "Empty {5}", y));
    if (x < 3)
      l.add(new Successor(new Water(3, y), "Fill up {3}", 3-x));
    if (y < 5)
      l.add(new Successor(new Water(x, 5), "Fill up {5}", 5-y));

    int a = 3-x;
    int b = 5-y;

    if (x < 3) {
      if (y >= a)
        l.add(new Successor(new Water(3, y-a), "Pour {5} -> {3}", a));
      else if (y > 0)
        l.add(new Successor(new Water(x+y, 0), "Pour {5} -> {3}", y));
    }
    if (y < 5) {
      if (x >= b)
        l.add(new Successor(new Water(x-b, 5), "Pour {3} -> {5}", b));
      else if (x > 0)
        l.add(new Successor(new Water(0, x+y), "Pour {3} -> {5}", x));
    }
    return l;
  }
}

File: source_code/python/water.py
class water:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return '[%d, %d]' % (self.x, self.y)

    def __repr__(self):
        return str(self)

    def isgoal(self):
        if self.y == 4:
            return True
        else:
            return False

    def successors(self):
        if self.x > 0:
            yield (water(0, self.y), "Empty {3}", self.x)
        if self.y > 0:
            yield (water(self.x, 0), "Empty {5}", self.y)

        if self.x < 3:
            yield (water(3, self.y), "Fill up {3}", 3 - self.x)
        if self.y < 5:
            yield (water(self.x, 5), "Fill up {5}", 5 - self.y)

        a = 3 - self.x
        b = 5 - self.y

        if self.x < 3:
            if self.y >= a:
                yield (water(3, self.y-a), "Pour {5} -> {3}", a)
            elif self.y > 0:
                yield (water(self.x+self.y, 0), "Pour {5} -> {3}", self.y)
        if self.y < 5:
            if self.x >= b:
                yield (water(self.x-b, 5), "Pour {3} -> {5}", b)
            elif self.x > 0:
                yield (water(0, self.x+self.y), "Pour {3} -> {5}", self.x)
